### PR TITLE
Document how stack uses the Cabal library in Architecture.md

### PR DESCRIPTION
**Pull request by dbaynard**
Relates to #3049, #4475, #4488.

Please include the following checklist in your PR:

* [x] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

### CODE REVIEW

**Comment by dbaynard** [comment]
The bulk of the content may belong in a different document.

File: doc/faq.md [hunk 1]
```diff
@@ -57,6 +57,65 @@ directory. None of this should affect any existing Haskell tools at all.
...
+
+The relationship between stack and Cabal-the-library is complicated.
+
+1. Stack itself builds against a version of the Cabal library, which it uses for parsing Cabal files.
+
+2. Stack wraps functionality from a version of the Cabal library in order to build packages.
+
+    i. For packages with `build-type: Simple`, Stack uses a setup executable compiled with a version of the Cabal library to perform builds. All packages using the same compiler and Cabal version are built with the same executable. These executables are cached in the `setup-exe-cache` configuration directory.
+
+    ii. For packages with `build-type: Custom`, Stack compiles `Setup.hs` against a version of the Cabal library and uses that setup executable to perform builds. The executable is stored TODO
```
Comment (snoyberg) [inline doc/faq.md]: "against a version of the Cabal library"- did you mean to include "specified by your stack.yaml file"?

File: doc/faq.md [hunk 2]
```diff
@@ -57,6 +57,65 @@ directory. None of this should affect any existing Haskell tools at all.
   there may be more options here for new projects in the future (see issue
   [253](https://github.com/commercialhaskell/stack/issues/253))
 
+### Cabal-the-library
+
+The relationship between stack and Cabal-the-library is complicated.
+
+1. Stack itself builds against a version of the Cabal library, which it uses for parsing Cabal files.
```
Comment (snoyberg) [inline doc/faq.md]: Worth pointing out that this determines which version of the cabal file format Stack is able to parse.

File: doc/faq.md [hunk 3]
```diff
@@ -57,6 +57,65 @@ directory. None of this should affect any existing Haskell tools at all.
...
+
+2. Stack wraps functionality from a version of the Cabal library in order to build packages.
```
Comment (snoyberg) [inline doc/faq.md]: "wraps" may be slightly confusing. Stack builds a `Setup.hs` file against a version of the Cabal library and then shells out to that executable. Stack itself does not perform any of the configure/build/copy/register stuff in-process. This allows it to more easily support newer versions of the Cabal library for performing builds, which is why _typically_ (but not always) you can use a Stack executable with a newer version of GHC than it was built for.

File: doc/faq.md [hunk 4]
```diff
@@ -57,6 +57,65 @@ directory. None of this should affect any existing Haskell tools at all.
...
+
+    ii. For packages with `build-type: Custom`, Stack compiles `Setup.hs` against a version of the Cabal library and uses that setup executable to perform builds. The executable is stored TODO
+
+3. Packages themselves may depend on the Cabal library.
+
+There are three (nearly) corresponding versions of the Cabal library which are relevant to a build.
+
+1. The version used to compile stack, which is not necessarily present on stack users’ machines.
+
+   Stack will always be compiled using the most recent stable version of Cabal-the-library.
```
Comment (snoyberg) [inline doc/faq.md]: Not really true. We may attempt to do that, but we could for example release a point release of an older Stack release without upgrading to the newest Cabal library version.

File: doc/faq.md [hunk 5]
```diff
@@ -57,6 +57,65 @@ directory. None of this should affect any existing Haskell tools at all.
...
+
+In addition, the Cabal version required by `custom-build` will be installed, if necessary.
+
+Note that these versions may be the same.
+
+There are a number of consequences of this design.
+
+1. Snapshot packages only depend on the GHC compiler version, not the Cabal library version (with the exception of `build-type: Custom` packages).
+
+2. The most recent stack can always read the most recent Cabal files.
```
Comment (snoyberg) [inline doc/faq.md]: "Usually"

File: doc/faq.md (hunk 1 above)
Comment (dbaynard) [reply doc/faq.md]: I think this is covered later — so it probably makes more sense to zip these two lists of numbered paragraphs (the relationship and the versions) and have each bit of functionality in a self-contained unit.

File: doc/faq.md (hunk 2 above)
Comment (dbaynard) [reply doc/faq.md]: ditto

File: doc/faq.md [hunk 6]
```diff
@@ -57,6 +57,65 @@ directory. None of this should affect any existing Haskell tools at all.
...
+      A library using `build-type: Simple` and SPDX license identifiers (introduced in Cabal 2.2) will not build, though stack will process the Cabal file correctly.
+
+          Cabal-simple_mPHDZzAJ_2.0.1.0_ghc-8.2.2: ./file.cabal:7: Parse of field 'license' failed.
+
+4. Stack only occasionally needs to build Cabal-the-library.
+   This is a resource intensive build, so avoiding it improves performance.
+
+5. Stack uses `ghc-pkg` to identify the Cabal version it should use for the setup executable for a build.
+   The build output is stored in a directory `.stack-work/dist/Cabal-xxxxx` named for the Cabal version used to compile the setup executable.
+   Together, this means it is only possible to know for sure which Cabal version ought to be used if the corresponding compiler is installed.
```
[the remark this answers is not in the document]
Comment (dbaynard) [reply doc/faq.md]: Ah — I didn't know that `global-hints.yaml` was new. What should we say here?

[the remark this answers is not in the document]
Comment (qrilka) [reply doc/faq.md]: As far as I understand the idea behind that file is to have information about boot packages without having GHC installed, currently it's used in (to be removed) `Stack.Snapshot` and in https://github.com/commercialhaskell/stack/blob/master/src/Stack/SourceMap.hs#L196 but that's actual for `master`, not for `stable`.

File: doc/faq.md (hunk 1 above)
Comment (dbaynard) [reply doc/faq.md]: Fixed at https://github.com/dbaynard/stack/blob/f31e9dd962aefa6c11854496ff578244618a9623/doc/faq.md#L85-L87

File: doc/faq.md (hunk 3 above)
Comment (dbaynard) [reply doc/faq.md]: I've used your language to fix.

File: doc/faq.md (hunk 4 above)
Comment (dbaynard) [reply doc/faq.md]: Added 'where possible' and made clear this refers to releases.

https://github.com/dbaynard/stack/blob/9900d99114967698223f43c84e4cedecc24cea97/doc/faq.md#L70-L72

**Comment by dbaynard** [comment]
Two commits:

1.  Address straightforward issues.
2.  Refactor completely to fix more complicated ones.

I still have a few things left to do, here (e.g. the TODO).

**Comment by dbaynard** [comment]
I need to put this in a new document, rather than the FAQ. Also, I'd like to rebase to take advantage of working CI on `stable`. I'll do all that when the content is OK.

**Comment by dbaynard** [comment]
I think this is ready, in terms of content (assuming the latest addition is correct; it is what my logs suggest though I haven't dug into the code). It should go in a new file, though, rather than the FAQ (perhaps with the 'What is the relationship between stack and cabal' bit directly above, too).

**Comment by dbaynard** [comment]
Actually, I moved the new material into `Architecture`, rather than a new document, and added links. I'll merge when it clears appveyor.

**Pull request merged**